cli/flox-rust-sdk/src/models/lockfile.rs

```diff
@@ -1,37 +1,198 @@
 pub struct Registry {
     pub inputs: BTreeMap<String, Input>,
     #[serde(flatten)]
     _json: Value,
 }
 
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct LockedManifest(Value);
+
+impl LockedManifest {
+    /// Use pkgdb to lock a manifest
+    pub fn lock_manifest(
+        pkgdb: &Path,
+        manifest_path: &Path,
+        existing_lockfile_path: Option<CanonicalPath>,
+        global_manifest_path: &Path,
+    ) -> Result<Self, LockedManifestError> {
+        let canonical_manifest_path = manifest_path
+            .canonicalize()
+            .map_err(|e| LockedManifestError::BadManifestPath(e, manifest_path.to_path_buf()))?;
+
+        let mut pkgdb_cmd = Command::new(pkgdb);
+        pkgdb_cmd
+            .args(["manifest", "lock"])
+            .arg("--ga-registry")
+            .arg("--global-manifest")
+            .arg(global_manifest_path)
+            .arg("--manifest")
+            .arg(canonical_manifest_path);
+        if let Some(canonical_lockfile_path) = existing_lockfile_path {
+            pkgdb_cmd.arg("--lockfile").arg(canonical_lockfile_path);
+        }
+
+        debug!("locking manifest with command: {pkgdb_cmd:?}");
+        call_pkgdb(pkgdb_cmd)
+            .map_err(LockedManifestError::LockManifest)
+            .map(Self)
+    }
+
+    /// Build a locked manifest
+    ///
+    /// if a gcroot_out_link_path is provided,
+    /// the environment will be linked to that path and a gcroot will be created
+    pub fn build(
+        &self,
+        pkgdb: &Path,
+        gcroot_out_link_path: Option<&Path>,
+    ) -> Result<PathBuf, LockedManifestError> {
+        let mut pkgdb_cmd = Command::new(pkgdb);
+        pkgdb_cmd.arg("buildenv").arg(&self.to_string());
+
+        if let Some(gcroot_out_link_path) = gcroot_out_link_path {
+            pkgdb_cmd.args(["--out-link", &gcroot_out_link_path.to_string_lossy()]);
+        }
+
+        debug!("building environment with command: {pkgdb_cmd:?}");
+
+        let result: BuildEnvResult =
+            serde_json::from_value(call_pkgdb(pkgdb_cmd).map_err(LockedManifestError::BuildEnv)?)
+                .map_err(LockedManifestError::ParseBuildEnvOutput)?;
+
+        Ok(PathBuf::from(result.store_path))
+    }
+}
+
+impl ToString for LockedManifest {
+    fn to_string(&self) -> String {
+        self.0.to_string()
+    }
+}
+
 /// An environment (or global) lockfile.
 ///
-/// The authoritative definition of this structure is in C++.
+/// This struct is meant **for reading only**.
+///
+/// It serves as a typed representation of the lockfile json produced by pkgdb.
+/// Parsing of the lockfile is done in [TypedLockedManifest::try_from]
+/// and should be as late as possible.
+/// Where possible, use the opaque [LockedManifest] instead of this struct
+/// to avoid incompatibility issues with the authoritative definition in C++.
+///
+/// In the optimal case the lockfile schema can be inferred from a common
```

Comment:
```suggestion
/// In the optimal case the lockfile schema can be inferred from a single
```

Reply:
I'm glad this comment identifies the risk.
I'm less enthusiastic that we're taking these risks.

I'd strongly prefer for this typed representation to be dropping in favor of inline access on raw JSON data.

Reply:
I'd prefer having the struct:
- It's more convenient to use
- If there is a change in schema, we can change it once and then rely on the compiler instead of grep
- We can essentially disallow modifying the lock by not deriving `Serialize`, whereas with a JSON `Value` that's less obvious

Can you explain what the advantage of inline access is? Or do you just feel having a struct is a slippery slope?

Reply:
The struct implies a complete typecheck of the entire record which is far more sensitive to changes than inline access to field names _as needed_.

If there's additional processing that needs to be performed on these records it belongs in C++.

Rust should not injest the lockfile or rely on its schema being stable.

I know we have a place today which already breaks this constraint by doing lockfile -> serde -> rust struct -> serde -> json in order to pretty print.
This risks loss of detail or fragility if schema changes.

If you need to access a field in the lockfile do so without a struct - use raw JSON datatype.
If you need to post process the lockfile - use C++.

Rust should write the lockfile as is with no modifications, or risk of modification in the event that schemas change.
If Rust cannot locate a field it may throw.
If it encounters an unexpected field it should ignore it.
If it encounters a `null` field, it should not remove it or interpret `null` with JSON-patch semantics ( delete the key/value ).

Reply:
Any writing operations go through `LockedManifest` all that is doing is sanity checking that we actually have a JSON document -- we can weaken that even more to just a String.

`TypedLockedManifest` is just a view into the JSON, extracting some fields.
That struct is also **only** used when we need to read fields from the lockfile in rust anyway to avoid us crawling to pkgdb to give a list opf packages (in this case) or extract the registry field.
For any other operation where we have a lockfile running through rust we just dump it without parsing it into anything.

regarding the typed lock parsing:
* extra fields are already ignored
* a field can't be located raises an error -- same as using raw json but with an actually useful error
* any extra information is lost or recovered as we chose
* we can protect ourselves from writing modified versions by a) enforcing private fields  on the struct b) providing no setters/modifiers and c) **not writing the struct** in the first place.

This all is not more fragile than using raw json access but consolidates changes to one place if the schema on the `pkgdb` side changes (which since its in the same codebase is mostly easy to detect and work against)